mpisppy/extensions/gradient_extension.py

```diff
@@ -40,60 +41,80 @@ def __init__(self, opt, comm=None):
         self.cfg.rho_path = './_temp_grad_rho_file.csv'
         self.grad_object = grad.Find_Grad(opt, self.cfg)
         self.rho_setter = find_rho.Set_Rho(self.cfg).rho_setter
-        self.primal_conv_cache = []
-        self.dual_conv_cache = []
+        self.prev_primal_norm, self.curr_primal_norm = 0, 0
         self.wt = WTracker(self.opt)
 
-    def _display_rho_values(self):
+
+## preliminary functions
+
+    def update_rho(self):
+        self.grad_object.write_grad_rho()
+        rho_setter_kwargs = self.opt.options['rho_setter_kwargs'] \
+                            if 'rho_setter_kwargs' in self.opt.options \
+                            else dict()
+        for sname, scenario in self.opt.local_scenarios.items():
+            rholist = self.rho_setter(scenario, **rho_setter_kwargs)
+            for (vid, rho) in rholist:
+                (ndn, i) = scenario._mpisppy_data.varid_to_nonant_index[vid]
+                scenario._mpisppy_model.rho[(ndn, i)] = rho
+        self.display_rho_values()
+
+
+    def _rho_primal_crit(self):
+        primal_thresh = 0.001 # or other value
+        self.prev_primal_norm = self.curr_primal_norm
+        self.curr_primal_norm = norms.scaled_primal_metric(self.opt)
+        norm_diff = np.abs(self.curr_primal_norm - self.prev_primal_norm)
+        #print(f'{norm_diff =}')
+        return (norm_diff <= primal_thresh)
+
+    def _rho_dual_crit(self):
+        dual_thresh = 0.1 # or other value
+        self.wt.grab_local_Ws()
+        dual_norm = norms.scaled_dual_metric(self.opt, self.wt.local_Ws, self.opt._PHIter)
```

Comment:
`_rho_primal_crit` is tracking changes in primal norm, but `_rho_dual_crit` is just tracking the absolute dual norm

Reply:
I updated the help for the config items for the thresholds.